New ContentView - Create, search, publish

latran opened the pull request:
Prelim work on the new content view with some basic functions.

Functions:
1. create new cv works
2. search a cv works
3. publish cv works to an extent
4. add repo works
5. update cv works

Issues:
**1. Need to implement some waiting period in entities for animation to finish loading, otherwise it loops the search. (using sleep for the time being)**
2. Writing PF4table.  I'm still having some issues with writing a robust PF4table
3. Need help discerning where to place the PF4table for `class NewContentViewEditView(BaseLoggedInView):`.  This is for adding repos.  The old table is placed in `common.py` for `class AddRemoveResourcesView(View):`.  I created a new one and put it in `class NewAddRemoveResourcesView(View):`.  My thinking is that this table will be common place in other future PF4 enhancements.
~~4. Getting when trying to access the different tabs in the table in editview: `{NoSuchElementException}Message: Could not find an element Locator(by='xpath', locator='.//ul[contains(@class, "nav-tabs")]/li[./a[normalize-space(.)="Versions"]]')`~~

Suggestions and feedback most welcomed to get me unstuck.

jameerpathan111 commented on airgun/views/new_contentview.py:
```diff
@@ -0,0 +1,82 @@
+    title = Text("//h1[contains(., 'Content Views')]")
+    create_content_view = PF4Button('OUIA-Generated-Button-primary-1')
+    #TODO: add expandable widget to PFTable
+    table = PatternflyTable(component_id='OUIA-Generated-Table-2')
```
I tried `table = PatternflyTable(component_id='OUIA-Generated-Table-2')` but that didn't work. The `component_id` of table was different after performing search operation, you will have to look for a better locator. You can try `table = PatternflyTable(locator='.//table[@data-ouia-component-type="PF4/Table"]')` but this one is also not that robust.

latran commented:
@jameerpathan111 @omkarkhatavkar @shweta83 @devendra104 

I've added a few more functionalities to the PR.

Issue 1 is still a big problem for me because there's a slight delay that airgun has that it can't recognize that it's in the correct place after navigating to it or searching for the cv/repos/etc.  You can see it in calls such as `view = self.navigate_to(self, 'Publish', entity_name=entity_name)` or `view = self.navigate_to(self, 'Edit', entity_name=entity_name)` or after searching for a cv/repo/etc.  I need help in how to fix this problem so that I don't have to add `time.sleep()` in areas where I know it can occur.  I've tried `wait_for` in some places and even `self.browser.wait_for_element(locator='//div[@class="pf-c-content"]')`.  Those seem to help for the time being, but I feel there should be a robust way of handling it.

The PR contains comments in sections where I'm having issues.  Please review and provide feedback so I can get unstuck.

omkarkhatavkar commented on airgun/views/new_contentview.py:
```diff
@@ -0,0 +1,174 @@
+    title = Text("//h1[contains(., 'Content Views')]")
+    create_content_view = PF4Button('OUIA-Generated-Button-primary-1')
+    table = PatternflyTable(
+        component_id='OUIA-Generated-Table-4',
```
I looked at the code in UI and it updates the component_id every time on the UI, we might need to inherit this PatternflyTable based on our need. I updated the underneath code and removed the component_id then the search seems to be working.

omkarkhatavkar replied:
Also looking at the data it retrieved it seems to be not correctly aligned in a dict. We need to work the right structure here.

latran replied:
@omkarkhatavkar what kind of update did you implement in the code besides removing the `component_id`?

latran commented:
@omkarkhatavkar 
A few updates.

I've removed `SearchableViewMixin` parameter in the views files because it seems we need to make a new SearchableViewMixin` specifically for PF4search it seems because it's using the old search.  So instead, I added the new search in the views file: `NewContentViewTableView`.  It seems to work without using `time.sleep()`
I also edited the `ouia.py` so the xpath is more flexible when reading the pf4 table.

```
locator=(
                f".//*[@data-ouia-component-type={quote(self.component_type)} "
                f"and contains(@data-ouia-component-id, {quote(self.component_id)})]"
            ),
```

latran commented:
Wrote a new `ProgressBarPF4` for publishing.  Added new `SearchableViewMixPF4` to utilize new search bar.  Added sleeper in places where I know automation needs time to load.  Next up is fixing PF4table so automation can read, promote environments, and create ccv.

@omkarkhatavkar Please review when you get a chance and maybe help with the search so we don't use sleepers.

latran commented:
Added several more view and functions.  In view, added copy & delete views, cv promote view, content_view tab for ccv, action dropdown widget, lce selection widget, and searching in versions.  also added function of delete, publish and promote, and promote. I have not tested this extensively yet, just adding these new views and functions in.

omkarkhatavkar commented on airgun/entities/new_contentview.py:
```diff
@@ -0,0 +1,225 @@
+        entity_name: name of content view
+    """
+
+    VIEW = NewContentViewRemoveView
```
This is missing view ? Cant locate in the views.

latran replied:
The renaming should fix this now.

omkarkhatavkar commented on airgun/views/new_contentview.py:
```diff
@@ -0,0 +1,262 @@
+    title = Text("//h1[contains(., 'Content Views')]")
+    create_content_view = PF4Button('OUIA-Generated-Button-primary-1')
+    table = PatternflyTable(
+        component_id='OUIA-Generated-Table',
```
Did Search works for you for me it did not ? I think we have to remove the `OUIA-Generated-Table'` inherit this table with our own table and add custom read()method. One way to do that is to self.read() after that remove the non required attributes.

omkarkhatavkar commented on airgun/widgets.py:
```diff
@@ -1830,6 +1881,71 @@ def read(self):
         }
 
 
+class ProgressBarPF4(GenericLocatorWidget):
```
here also if you can able to use same implementation from `ProgressBar` class inherit that

latran commented:
Added the new changes from omkar that does reading of the content view table in 'all'.

omkarkhatavkar commented on setup.py:
```diff
@@ -21,12 +21,11 @@
         'widgetastic.core',
         'widgetastic.patternfly',
-        'widgetastic.patternfly4',
+        'widgetastic.patternfly4 @ git+https://github.com/RedHatQE/widgetastic.patternfly4.git@selenium4#egg=widgetastic.patternfly4',  # noqa
```
looks like need to rebase with latest master

omkarkhatavkar commented:
Can one of the admins verify this patch?

jyejare commented:
Ping @Griffin-Sullivan , rebase the PR based on what we have in master.

sambible commented:
Closing this PR, as these changes have been picked up and completed in other, more recent Airgun PRs.
